kura/org.eclipse.kura.asset.provider/src/main/java/org/eclipse/kura/asset/provider/DriverState.java

```diff
@@ -84,46 +88,50 @@ public void syncChannelListeners(final Set<ChannelListenerRegistration> targetSt
     private void setChannelListenersInternal(final Set<ChannelListenerRegistration> targetState,
             final Map<String, Channel> channels) {
 
-        final Iterator<ChannelListenerRegistration> iter = this.attachedListeners.iterator();
-
-        while (iter.hasNext()) {
-            final ChannelListenerRegistration reg = iter.next();
-
-            if (!targetState.contains(reg)) {
-                detach(iter, reg);
-            }
+        List<ChannelListenerRegistration> tobeRemoved = this.attachedListeners.stream()
+                .filter(reg -> !targetState.contains(reg)).collect(Collectors.toList());
+        if (!tobeRemoved.isEmpty()) {
+            this.detach(tobeRemoved);
         }
 
-        for (final ChannelListenerRegistration reg : targetState) {
-
+        Map<Channel, ChannelListenerRegistration> regChannels = targetState.stream().filter(reg -> {
             if (this.attachedListeners.contains(reg)) {
-                continue;
+                return false;
             }
 
             final Channel channel = channels.get(reg.getChannelName());
 
-            if (channel != null && channel.isEnabled()) {
-                attach(reg, channel);
-            }
+            return channel != null && channel.isEnabled();
+        }).collect(Collectors.toMap(reg -> channels.get(reg.getChannelName()), reg -> reg));
+        if (!regChannels.isEmpty()) {
+            attach(regChannels);
         }
+
     }
 
-    private void attach(final ChannelListenerRegistration reg, final Channel channel) {
+    private void attach(final Map<Channel, ChannelListenerRegistration> regChannels) {
         try {
             logger.debug("Registering Channel Listener for monitoring...");
-            this.driver.registerChannelListener(channel.getConfiguration(), reg.getChannelListener());
-            this.attachedListeners.add(reg);
```

Comment:
Hi @spnettec, thank you for the contribution. The current implementation of the Asset handles the case of a failure in registering a single channel listener, and if this happens the Asset will attempt to register other listen channels in its configuration (if any). Is this behavior preserved in the new implementation?

Reply:
I would like to confirm that based on the results of the unit test is it possible to register multiple listener instances for one channel Instead of a channel only one instance of the listener can be registered?  At present what I have modified is that a channel can only listen to one listener instance.